// react-carousel-cards-npm/dist/esm/components/SideCard.js
import { __assign } from "tslib";
import React from 'react';
import stylesCss from '../styles/carousel.module.css';
import { renderChildren } from '../utils';
var SideCard = function (_a) {
    var child = _a.child, cardWidth = _a.cardWidth, item = _a.item, width = _a.width, index = _a.index, selected = _a.selected, widthCard = _a.widthCard, _b = _a.marginCard, marginCard = _b === void 0 ? 1 : _b;
    var marginRight = width < 900 ? marginCard / 2 : marginCard * (selected - 1);
    var marginLeft = width < 900 ? marginCard / 2 : marginCard;
    return (React.createElement("div", { className: stylesCss['side-card-box'], style: { right: index === 0 ? 'auto' : 0 }, key: item.key },
        renderChildren(child, __assign(__assign(__assign({}, item), styles.card), { maxWidth: cardWidth, style: styles.sideCardParent })),
        renderChildren(child, __assign(__assign({}, item), { style: __assign(__assign(__assign({}, styles.sideCardChild), styles.card), { maxWidth: cardWidth, right: index === 0
                    ? "calc(100% + ".concat(marginLeft, "px")
                    : width - (selected - 1) * widthCard - widthCard - marginRight + 'px' }) }))));
};
export default SideCard;
var styles = {
    card: {
        width: '100%',
    },
    sideCardParent: {
        width: '30%',
        height: '0',
        overflow: 'hidden',
        border: 'none',
    },
    sideCardChild: {
        position: 'absolute',
        top: 0,
    },
};
