// react-carousel-cards-npm/dist/cjs/hooks/useCaruselContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useCarouselContext = exports.CarouselContextProvider = exports.CarouselContext = void 0;
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importStar(require("react"));
var utils_1 = require("../utils");
var defaultContext = {
    totalPageCount: 1,
    handleNextPage: function () { },
    handlePrevPage: function () { },
    onCurrentPage: function () { },
    onTotalPageCountChange: function () { },
};
exports.CarouselContext = (0, react_1.createContext)(defaultContext);
var CarouselContextProvider = function (_a) {
    var children = _a.children;
    var _b = (0, react_1.useState)(1), currentPage = _b[0], onCurrentPage = _b[1];
    var _c = (0, react_1.useState)(1), totalPageCount = _c[0], onTotalPageCountChange = _c[1];
    var handleNextPage = (0, utils_1.setNextPage)(onCurrentPage);
    var handlePrevPage = (0, utils_1.setPrevPage)(onCurrentPage);
    return (react_1.default.createElement(exports.CarouselContext.Provider, { value: { currentPage: currentPage, onCurrentPage: onCurrentPage, handlePrevPage: handlePrevPage, handleNextPage: handleNextPage, totalPageCount: totalPageCount, onTotalPageCountChange: onTotalPageCountChange } }, children));
};
exports.CarouselContextProvider = CarouselContextProvider;
var useCarouselContext = function () {
    return (0, react_1.useContext)(exports.CarouselContext);
};
exports.useCarouselContext = useCarouselContext;
