// react-carousel-cards-npm/dist/esm/hooks/useCaruselContext.js
import React, { createContext, useContext, useState } from 'react';
import { setNextPage, setPrevPage } from '../utils';
var defaultContext = {
    totalPageCount: 1,
    handleNextPage: function () { },
    handlePrevPage: function () { },
    onCurrentPage: function () { },
    onTotalPageCountChange: function () { },
};
export var CarouselContext = createContext(defaultContext);
export var CarouselContextProvider = function (_a) {
    var children = _a.children;
    var _b = useState(1), currentPage = _b[0], onCurrentPage = _b[1];
    var _c = useState(1), totalPageCount = _c[0], onTotalPageCountChange = _c[1];
    var handleNextPage = setNextPage(onCurrentPage);
    var handlePrevPage = setPrevPage(onCurrentPage);
    return (React.createElement(CarouselContext.Provider, { value: { currentPage: currentPage, onCurrentPage: onCurrentPage, handlePrevPage: handlePrevPage, handleNextPage: handleNextPage, totalPageCount: totalPageCount, onTotalPageCountChange: onTotalPageCountChange } }, children));
};
export var useCarouselContext = function () {
    return useContext(CarouselContext);
};
