// react-carousel-cards-npm/dist/cjs/components/Carousel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importStar(require("react"));
var useCarusel_1 = require("../hooks/useCarusel");
var useResize_1 = require("../hooks/useResize");
var utils_1 = require("../utils");
var carousel_module_css_1 = tslib_1.__importDefault(require("../styles/carousel.module.css"));
var SideCard_1 = tslib_1.__importDefault(require("./SideCard"));
var CustomArrowsWrapper_1 = tslib_1.__importDefault(require("./CustomArrowsWrapper"));
var useCaruselContext_1 = require("../hooks/useCaruselContext");
var useAnimation_1 = require("../hooks/useAnimation");
var Carousel = function (_a) {
    var i18n = _a.i18n, defaultActivePage = _a.defaultActivePage, defaultCardsCount = _a.defaultCardsCount, _b = _a.variant, variant = _b === void 0 ? 'regular' : _b, disabled = _a.disabled, _c = _a.cards, cards = _c === void 0 ? [] : _c, _d = _a.cardWidth, cardWidth = _d === void 0 ? 300 : _d, cardContainerStyles = _a.cardContainerStyles, carouselContainerStyles = _a.carouselContainerStyles, paginationButtonStyles = _a.paginationButtonStyles, _e = _a.marginCard, marginCard = _e === void 0 ? 1 : _e, header = _a.header, _f = _a.noCardsText, noCardsText = _f === void 0 ? 'No cards selected' : _f, children = _a.children, CustomArrowBtn = _a.CustomArrowBtn, CustomPaginationBtn = _a.CustomPaginationBtn, CustomNoCardsBlock = _a.CustomNoCardsBlock, withAnimation = _a.withAnimation;
    var _g = (0, react_1.useState)(defaultCardsCount || 3), selected = _g[0], setSelected = _g[1];
    var _h = (0, react_1.useState)(0), touchStart = _h[0], setTouchStart = _h[1];
    var _j = (0, react_1.useState)(0), touchEnd = _j[0], setTouchEnd = _j[1];
    var _k = (0, react_1.useState)(1), currentPageLocal = _k[0], setCurrentPageLocal = _k[1];
    var ref = (0, react_1.useRef)(null);
    var refCard = (0, react_1.useRef)(null);
    var _l = (0, useCaruselContext_1.useCarouselContext)(), contextCurrentPage = _l.currentPage, contextOnCurrentPage = _l.onCurrentPage, contextHandlePrevPage = _l.handlePrevPage, contextHandleNextPage = _l.handleNextPage, onTotalPageCountChange = _l.onTotalPageCountChange;
    var currentPage = contextCurrentPage || 1;
    var handleNextPage = contextHandleNextPage;
    var onCurrentPage = contextOnCurrentPage;
    var handlePrevPage = contextHandlePrevPage;
    if (!contextCurrentPage) {
        currentPage = currentPageLocal;
        handleNextPage = (0, utils_1.setNextPage)(setCurrentPageLocal);
        onCurrentPage = setCurrentPageLocal;
        handlePrevPage = (0, utils_1.setPrevPage)(setCurrentPageLocal);
    }
    var isPaginationShown = variant !== 'withoutPagination' && !variant.includes('withoutPagination');
    var isArrowsShown = variant !== 'withoutArrows' && !variant.includes('withoutArrows');
    var isSideCardsShown = variant === 'withSideCards' || variant.includes('withSideCards');
    var isRegularCardsShown = variant === 'regular' || variant.includes('regular');
    var _m = (0, useCarusel_1.useCarousel)({
        selected: selected,
        cards: cards,
        currentPage: currentPage,
        isSideCardsShown: isSideCardsShown,
    }), totalPageCount = _m.totalPageCount, selectedCards = _m.selectedCards;
    var _o = (0, useResize_1.useResize)({
        ref: ref,
        setSelected: setSelected,
        cardWidth: cardWidth,
        refCard: refCard,
        isRegularCardsShown: isRegularCardsShown,
    }), width = _o.width, widthCard = _o.widthCard;
    var isAnimate = (0, useAnimation_1.useAnimation)({ withAnimation: withAnimation, selectedCards: selectedCards }).isAnimate;
    var handleTouchStart = function (e) { return setTouchStart(e.targetTouches[0].clientX); };
    var handleTouchMove = function (e) { return setTouchEnd(e.targetTouches[0].clientX); };
    var handleTouchEnd = function () {
        if (touchStart - touchEnd > 10) {
            /** Swipe Right **/
            if (currentPage < totalPageCount) {
                handleNextPage();
            }
        }
        if (touchStart - touchEnd < -10) {
            /** Swipe Left **/
            if (currentPage > 1) {
                handlePrevPage();
            }
        }
    };
    var keyDownHandle = function (e) {
        /** Key arrow Right **/
        if (e.key === 'ArrowRight') {
            if (currentPage < totalPageCount) {
                handleNextPage();
            }
        }
        /** Key arrow Left **/
        if (e.key === 'ArrowLeft') {
            handlePrevPage();
        }
    };
    var paginationHandle = function (index) {
        onCurrentPage(index + 1);
    };
    (0, react_1.useEffect)(function () {
        if (defaultActivePage) {
            onCurrentPage(defaultActivePage);
        }
    }, [defaultActivePage, onCurrentPage]);
    (0, react_1.useEffect)(function () {
        onTotalPageCountChange(totalPageCount);
    }, [totalPageCount, onTotalPageCountChange]);
    (0, react_1.useEffect)(function () {
        if (currentPage > totalPageCount) {
            setCurrentPageLocal(totalPageCount);
            contextOnCurrentPage(totalPageCount);
        }
    }, [totalPageCount, currentPage]);
    (0, react_1.useEffect)(function () {
        window.addEventListener('keydown', keyDownHandle);
        return function () { return window.removeEventListener('keydown', keyDownHandle); };
    }, [keyDownHandle]);
    return (react_1.default.createElement("div", { className: carousel_module_css_1.default['carousel-container'], ref: ref, style: carouselContainerStyles, "data-testid": "carousel-".concat(i18n) },
        react_1.default.createElement("div", { className: carousel_module_css_1.default['carousel-container__header'] },
            header,
            isArrowsShown && totalPageCount > 1 && (react_1.default.createElement(CustomArrowsWrapper_1.default, { CustomArrowBtn: CustomArrowBtn, currentPage: currentPage, lastPage: totalPageCount, disabled: disabled, handleNextPage: handleNextPage, handlePrevPage: handlePrevPage }))),
        react_1.default.createElement("div", { className: carousel_module_css_1.default['carousel-container__body'], "data-testid": 'carousel-container__body', style: cardContainerStyles, onTouchStart: handleTouchStart, onTouchMove: handleTouchMove, onTouchEnd: handleTouchEnd },
            react_1.default.createElement("div", { ref: refCard, style: { height: 0, width: '100%', maxWidth: cardWidth + 'px' } }),
            react_1.default.createElement("div", { style: tslib_1.__assign(tslib_1.__assign(tslib_1.__assign({ gap: marginCard + 'px', display: 'flex', width: '100%' }, (withAnimation && {
                    opacity: isAnimate ? '1' : '0',
                    transition: 'opacity 0.5s ease-in-out',
                })), ((withAnimation === null || withAnimation === void 0 ? void 0 : withAnimation.animateIn) && isAnimate ? withAnimation.animateIn : withAnimation === null || withAnimation === void 0 ? void 0 : withAnimation.animateOut)), cardContainerStyles) }, !cards.length ? (react_1.default.createElement(react_1.default.Fragment, null, CustomNoCardsBlock ? ((0, utils_1.renderChild)(CustomNoCardsBlock, {
                noCardsText: noCardsText,
            })) : (react_1.default.createElement("div", { className: carousel_module_css_1.default['no-cards-container'] },
                react_1.default.createElement("span", { className: carousel_module_css_1.default['carousel-container__text'] }, noCardsText))))) : (selectedCards.map(function (item, index) {
                if (!isAnimate && withAnimation) {
                    return (0, utils_1.renderChildren)(children, tslib_1.__assign(tslib_1.__assign({}, item), { style: {
                            maxWidth: cardWidth,
                            visibility: 'hidden',
                        } }));
                }
                if (isSideCardsShown) {
                    if ((index === 0 && currentPage !== 1) ||
                        (selected < selectedCards.length && index === selectedCards.length - 1) ||
                        (index === selected - 1 && currentPage === 1)) {
                        return (react_1.default.createElement(SideCard_1.default, { index: index, item: item, width: width, key: item.key, selected: selected, cardWidth: cardWidth, widthCard: widthCard, marginCard: marginCard, child: children }));
                    }
                }
                return (0, utils_1.renderChildren)(children, tslib_1.__assign(tslib_1.__assign({}, item), { style: tslib_1.__assign(tslib_1.__assign({}, styles.card), { maxWidth: cardWidth }) }));
            })))),
        isPaginationShown && totalPageCount > 1 && (react_1.default.createElement("div", { className: carousel_module_css_1.default['carousel-container__pagination-box'] }, tslib_1.__spreadArray([], cards, true).splice(0, totalPageCount).map(function (item, index) {
            return CustomPaginationBtn ? ((0, utils_1.renderChild)(CustomPaginationBtn, {
                onClick: function () { return paginationHandle(index); },
                key: "".concat(item.key, "-button"),
                disabled: disabled,
                isActivePage: currentPage === index + 1,
            })) : (react_1.default.createElement("button", { className: carousel_module_css_1.default['pagination-button'], disabled: disabled, key: "".concat(item.key, "-button"), onClick: function () { return paginationHandle(index); }, style: tslib_1.__assign(tslib_1.__assign({}, paginationButtonStyles), { cursor: disabled ? 'not-allowed' : 'pointer', backgroundColor: currentPage === index + 1 ? 'hsla(200, 85%, 60%, 0.8)' : 'rgba(0, 0, 0, 0.1)' }), "data-testid": "pagination-Button".concat(index + 1) }));
        })))));
};
exports.default = Carousel;
var styles = {
    card: {
        width: '100%',
    },
};
