// react-carousel-cards-npm/dist/cjs/components/SideCard.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var carousel_module_css_1 = tslib_1.__importDefault(require("../styles/carousel.module.css"));
var utils_1 = require("../utils");
var SideCard = function (_a) {
    var child = _a.child, cardWidth = _a.cardWidth, item = _a.item, width = _a.width, index = _a.index, selected = _a.selected, widthCard = _a.widthCard, _b = _a.marginCard, marginCard = _b === void 0 ? 1 : _b;
    var marginRight = width < 900 ? marginCard / 2 : marginCard * (selected - 1);
    var marginLeft = width < 900 ? marginCard / 2 : marginCard;
    return (react_1.default.createElement("div", { className: carousel_module_css_1.default['side-card-box'], style: { right: index === 0 ? 'auto' : 0 }, key: item.key },
        (0, utils_1.renderChildren)(child, tslib_1.__assign(tslib_1.__assign(tslib_1.__assign({}, item), styles.card), { maxWidth: cardWidth, style: styles.sideCardParent })),
        (0, utils_1.renderChildren)(child, tslib_1.__assign(tslib_1.__assign({}, item), { style: tslib_1.__assign(tslib_1.__assign(tslib_1.__assign({}, styles.sideCardChild), styles.card), { maxWidth: cardWidth, right: index === 0
                    ? "calc(100% + ".concat(marginLeft, "px")
                    : width - (selected - 1) * widthCard - widthCard - marginRight + 'px' }) }))));
};
exports.default = SideCard;
var styles = {
    card: {
        width: '100%',
    },
    sideCardParent: {
        width: '30%',
        height: '0',
        overflow: 'hidden',
        border: 'none',
    },
    sideCardChild: {
        position: 'absolute',
        top: 0,
    },
};
